[ZIP 312] FROST for Spend Authorization Signatures (#662)
zcash/zips: zip-0312.rst

::

  ZIP: 312

>   Title: FROST for Spend Authorization Signatures

Reviewer: ```suggestion
  Title: FROST for Spend Authorization Multisignatures
```
"FROST for Spend Authorization Signatures" doesn't tell you what it's for unless you already know what FROST is for.

following modifications.

Key Generation
''''''''''''''

While key generation is out of scope for this ZIP and the FROST spec [#FROST]_,
it needs to be consistent with FROST, see [#frost-tdkg]_ for guidance. The spend
authorization private key :math:`\mathsf{ask}` [#protocol-spendauthsig]_ is the
particular key that must be used in the context of this ZIP. Note that the
:math:`\mathsf{ask}` is usually derived from the spending key
:math:`\mathsf{sk}`, though that is not required. Not doing so allows using
distributed key generation, since the key it generates is unpredictable. Note
however that not deriving :math:`\mathsf{ask}` from :math:`\mathsf{sk}` prevents
using seed phrases to recover the original secret (which may be something
desirable in the context of FROST).

Randomizer Generation
'''''''''''''''''''''

A new helper function is defined, which generates a randomizer. The
`encode_signing_package` is defined as the byte serialization of the `msg`,
`commitment_list` values as described in [#frost-serialization]_.
Implementations MAY choose another encoding as long as all values (the message,
and the identifier, binding nonce and hiding nonce for each participant) are

> encoded.

Author: ```suggestion
unambiguously encoded.
```

::

  randomizer_generate():

  Inputs:
  - msg, the message being signed in the current FROST signing run
  - commitment_list = [(i, hiding_nonce_commitment_i,
    binding_nonce_commitment_i), ...], a list of commitments issued by
    each participant, where each element in the list indicates a
    NonZeroScalar identifier i and two commitment Element values
    (hiding_nonce_commitment_i, binding_nonce_commitment_i). This list
    MUST be sorted in ascending order by identifier.

  Outputs: randomizer, a Scalar

  def randomizer_generate(msg, commitment_list):
    rng_randomizer = G.RandomScalar()

>     rng_randomizer_enc = G.SerializeScalar(rng_randomizer)

Author: Change to random byte buffer

Author: Done in 2018579

    signing_package_enc = encode_signing_package(commitment_list, msg)

>     randomizer_input = random_bytes(rng_randomizer_enc || signing_package_enc)

Author: ```suggestion
    randomizer_input = rng_randomizer_enc || signing_package_enc
```

- Hash (``H``): BLAKE2b-512 [#BLAKE]_ (BLAKE2b with 512-bit output and 16-byte personalization string),
  and Nh = 64.

  - H1(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubR", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo ``G.Order()``.
  - H2(m): Implemented by computing BLAKE2b-512("Zcash_RedJubjubH", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo ``G.Order()``.
    (This is equivalent to :math:`\mathsf{H}^\circledast(m)`, as defined by
    the :math:`\mathsf{RedJubjub}` scheme instantiated in [#protocol-concretereddsa]_.)
  - H3(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubN", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo ``G.Order()``.
  - H4(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubM", m).
  - H5(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubC", m).
  - HR(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubA", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer

>     modulo ``G.Order()``.

Author: Check if some of those can be matched to RedDSA hashes

Author: RedDSA basically uses the same function for HR and H2. I don't think we gain anything by making them them equal in this ZIP since FROST already requires a bunch of domain-separated hash functions anyway.

- Like the FROST specification [#FROST]_, this ZIP does not specify a key generation
  procedure; but refer to that specification for guidelines.
- Network privacy is not in scope for this ZIP, and must be obtained with other
  tools if desired.

Specification
=============

Algorithms in this section are specified using Python pseudo-code, in the same
fashion as the FROST specification [#FROST]_.

The types Scalar, Element, and G are defined in [#frost-primeordergroup]_, as
well as the notation for elliptic-curve arithmetic, which uses the additive
notation. Note that this notation differs from that used in the Zcash Protocol
Specification. For example, ``G.ScalarMult(P, k)`` is used for scalar
multiplication, where the protocol spec would use :math:`[k] P` with the group
implied by :math:`P`.

An additional per-ciphersuite hash function is used, denote ``HR(m)``, which
receives an arbitrary-sized byte string and returns a Scalar. It is defined
concretely in the Ciphersuites section.

Re-randomizable FROST
---------------------

To add re-randomization to FROST, follow the specification [#FROST]_ with the
following modifications.

Key Generation

> ''''''''''''''

Reviewer 2: Move this subsection to instead be a section before "Re-randomizable FROST". Given that FROST does not specify key generation, it doesn't make sense to define this section as a modification to the FROST specification.

Author: Done in 6e6893d8323103013163586eb7e3c53e929a14b4

> While key generation is out of scope for this ZIP and the FROST spec [#FROST]_,
> it needs to be consistent with FROST, see [#frost-tdkg]_ for guidance. The spend
> authorization private key :math:`\mathsf{ask}` [#protocol-spendauthsig]_ is the
> particular key that must be used in the context of this ZIP. Note that the
> :math:`\mathsf{ask}` is usually derived from the spending key
> :math:`\mathsf{sk}`, though that is not required. Not doing so allows using
> distributed key generation, since the key it generates is unpredictable. Note
> however that not deriving :math:`\mathsf{ask}` from :math:`\mathsf{sk}` prevents
> using seed phrases to recover the original secret (which may be something
> desirable in the context of FROST).

Reviewer 2: ```suggestion
While key generation is out of scope for this ZIP and the FROST spec [#FROST]_,
it needs to be consistent with FROST; see [#frost-tdkg]_ for general guidance.

To define a spending or viewing key that uses FROST, the Sapling and Orchard key
trees [#protocol-saplingkeycomponents]_ [#protocol-orchardkeycomponents]_ are
adjusted as follows:

- The Spend validating key :math:`\mathsf{ak}` is replaced by the FROST group
  public key `PK` [#frost-protocol]_.
- The Spend authorizing key :math:`\mathsf{ask}` is replaced by the logical
  signing key that corresponds to the group public key `PK`. By design, this
  key never exists, and instead is represented by each participant's FROST
  signing key share `sk_i`.

The remaining parts of the Sapling and Orchard key trees are generated from
a common... (TODO: Finish specifying how the other common parts of the
Sapling and Orchard key trees are derived for participants, perhaps in terms
of a common `sk` or a common HD path.)

(Old remaining content below, which might change after the above TODO.)
Note that the :math:`\mathsf{ask}` is usually derived from the spending key
:math:`\mathsf{sk}`, though that is not required. Not doing so allows using
distributed key generation, since the key it generates is unpredictable. Note
however that not deriving :math:`\mathsf{ask}` from :math:`\mathsf{sk}` prevents
using seed phrases to recover the original secret (which may be something
desirable in the context of FROST).
```

Reviewer 2: Moved to #883.

::

  ZIP: 312
  Title: FROST for Spend Authorization Signatures
  Owners: Conrado Gouvea <conrado@zfnd.org>
          Chelsea Komlo <ckomlo@uwaterloo.ca>
          Deirdre Connolly <deirdre@zfnd.org>
  Status: Draft
  Category: Wallet
  Created: 2022-08-dd
  License: MIT
  Discussions-To: <https://github.com/zcash/zips/issues/382>

>   Pull-Request: <https://github.com/zcash/zips/pull/TODO>

Reviewer: ```suggestion
  Pull-Request: <https://github.com/zcash/zips/pull/662>
```

signatures in the Zcash protocol, for the Sapling and Orchard network upgrades.

Motivation
==========

In the Zcash protocol, Spend Authorization Signatures are employed to authorize
a transaction. The ability to generate these signatures with the user's
private key is what effectively allows the user to spend funds.

This is a security-critical step, since anyone who obtains access to the private
key will be able to spend the user's funds. For this reason, one interesting
possibility is to require multiple parties to allow the transaction to go through.
This can be accomplished with threshold signatures, where the private key is
split between parties in a way that a threshold (e.g. 2 out of 3) of them must
sign the transaction in order to create the final signature. This enables scenarios
such as users and exchanges sharing custody of a wallet, for example.

FROST is one of such threshold signature protocols. However, it can't be used as-is
since the Zcash protocol also requires re-randomizing public and private keys
to ensure unlinkability between transactions. This ZIP specifies a variant of
FROST with re-randomization support.

Requirements
============

> - All signatures generated by following this ZIP must be verified successfully.
>   as Sapling or Orchard spend authorization signatures.

Reviewer: ```suggestion
- All signatures generated by following this ZIP must be verified successfully
  as Sapling or Orchard spend authorization signatures using the appropriate
  validating key.
```

since the Zcash protocol also requires re-randomizing public and private keys
to ensure unlinkability between transactions. This ZIP specifies a variant of
FROST with re-randomization support.

Requirements
============

- All signatures generated by following this ZIP must be verified successfully
  as Sapling or Orchard spend authorization signatures using the appropriate
  validating key.
- The signatures generated by following this ZIP should meet the security criteria
  for Signature with Re-Randomizable Keys as specified in the Zcash protocol [#protocol-concretereddsa]_.
- The threat model described below must be taken into account.

Threat Model
------------

In normal usage, a Zcash user follows multiple steps in order to generate a
shielded transaction:

- The transaction is created.
- The transaction is signed with a re-randomized version of the user's spend
  authorization private key.
- The zero-knowledge proof for the transaction is created with the randomizer
  as an auxiliary (secret) input, among others.

When employing re-randomizable FROST as specified in this ZIP, the goal is to
split the spend authorization private key :math:`\mathsf{ask}` among multiple

> possible signers. This means that the proof generation will still be generated

Reviewer: ```suggestion
possible signers. This means that the proof generation will still be performed
```

by a single participant, likely the one that created the transaction in the first
place. Note that this user already controls the privacy of the transaction since
they are responsible for creating the proof.

This fits well into the "Coordinator" role from the FROST specification [#frost-protocol]_.
The Coordinator is responsible for sending the message to be signed to all participants,
and to aggregate the signature shares.

With those considerations in mind, the threat model considered in this ZIP is:

- The Coordinator is trusted with the privacy of the transaction (which includes
  the unlinkability property). A rogue Coordinator will be able to break
  unlinkability and privacy, but should not be able to create signed transactions
  without the approval of `MIN_SIGNERS` participants, as specified in FROST.

> - All key share holders are also trusted with the privacy and of the transaction,

Reviewer: ```suggestion
- All key share holders are also trusted with the privacy of the transaction,
```

  thus a rogue key share holder will be able to break its privacy and unlinkability.
  A future specification may support a scenario where individual key share

>   holders are not trusted with it.

Reviewer: ```suggestion
  holders are not trusted with privacy or unlinkability.
```
Is this possible? Each of the key share holders need to see any transaction they are contributing to signing (less authorizing data, and probably with at least some of the commitments/ciphertexts opened/decrypted so that they can check amounts and recipients), in order to decide whether or not to sign it. Therefore any of them can decide to breach unlinkability by recognizing those transactions on the block chain and publishing that they are linked.

Reviewer: I think that if a participant were to publish everything they were sent and their randomness, this would be necessarily be sufficient for a proof (at least, absent changes that we have no idea how to make).

Non-requirements
================

- This ZIP does not support removing the Coordinator role, as described in #[frost-removingcoordinator]_.
- This ZIP does not prevent key share holders from linking the signing operation to a transaction in the blockchain.
- Like the FROST specification [#FROST], this ZIP does not specify a key generation
  procedure; but refer to that specification for guidelines.
- Network privacy is not in scope for this ZIP, and must be obtained with other
  tools if desired.

Specification
=============

Algorithms in this section are specified using Python pseudo-code, in the same
fashion as the FROST specification [#FROST]_.

The types Scalar, Element, and G are defined in #[frost-primeordergroup]_, as well

> as the notation for elliptic-curve arithmetic, which uses the additive notation.

Reviewer: ```suggestion
as the additive notation used for elliptic-curve arithmetic.
```

Re-randomizable FROST
---------------------

To add re-randomization to FROST, follow the specification [#FROST]_ with the
following modifications.

Key Generation
''''''''''''''

While key generation is out of scope for this ZIP and the FROST spec [#FROST]_,
it needs to be consistent with FROST, see [#frost-tdkg]_ for guidance. The
spend authorization private key :math:`\mathsf{ask}` [#protocol-spendauthsig]_
is the particular key that must be used in the context of this ZIP. Note that
while Sapling allows creating it directly, in Orchard it is always derived

> from the spending key :math:`\mathsf{ask}`. This means that a trusted

Reviewer: $\mathsf{ask}$ is derived from $\mathsf{ask}$? That sounds wrong.

Reviewer: It's not a consensus requirement of either Sapling or Orchard that  be generated in any particular way. In practice it is generated according to ZIP 32 in the single-signer case, but that doesn't really constrain how it is generated in the multi-signer case.

dealer key generation process might be required as detailed in [#frost-tdkg]_.

Randomizer Generation
'''''''''''''''''''''

A new helper function is defined, which computes :math:`\mathsf{RedDSA.GenRandom}`:

::

  randomizer_generate():

  Inputs:
  - None

  Outputs: randomizer, a Scalar

  def randomizer_generate():
    randomizer_input = random_bytes(64)
    return H3(randomizer_input)

Binding Factor Computation
''''''''''''''''''''''''''

> The `compute_binding_factor` function is changed to receive the `randomizer_point`

Reviewer: ```suggestion
The `compute_binding_factors` function is changed to receive the `randomizer_point`
```

as follows: ::

  Inputs:
  - commitment_list = [(i, hiding_nonce_commitment_i, binding_nonce_commitment_i), ...],
    a list of commitments issued by each participant, where each element in the list
    indicates a NonZeroScalar identifier i and two commitment Element values
    (hiding_nonce_commitment_i, binding_nonce_commitment_i). This list MUST be sorted
    in ascending order by identifier.
  - msg, the message to be signed.
  - randomizer_point, an element in G.

  Outputs:
  - binding_factor_list, a list of (NonZeroScalar, Scalar) tuples representing the binding factors.

  def compute_binding_factors(commitment_list, msg, randomizer_point):
    msg_hash = H4(msg)
    encoded_commitment_hash = H5(encode_group_commitment_list(commitment_list))
    rho_input_prefix = msg_hash || encoded_commitment_hash

    binding_factor_list = []
    for (identifier, hiding_nonce_commitment, binding_nonce_commitment) in commitment_list:
      rho_input = rho_input_prefix || G.SerializeScalar(identifier)
      binding_factor = H1(rho_input)
      binding_factor_list.append((identifier, binding_factor))

>     return binding_factor_list

Reviewer: This does not use `randomizer_point`. Was it intended to pass an encoding of it to H5 or include it in `rho_input_prefix`?

Round One - Commitment
''''''''''''''''''''''

Roune One is exactly the same as specified #[FROST]_. But for context, it
involves these steps:

- Each signer generates nonces and their corresponding public commitments.
  A nonce is a pair of Scalar values, and a commitment is a pair of Element values.
- The nonces are stored locally by the signer and kept private for use in the second round.
- The commitments are sent to the Coordinator.

Round Two - Signature Share Generation
''''''''''''''''''''''''''''''''''''''

In Round Two, the Coordinator generates a random scalar `randomizer` by calling
`randomizer_generate`. Then it computes `randomizer_point = G.ScalarBaseMult(randomizer)`
and sends it to each signer, over a confidential and authenticated channel,
along with the message and the set of signing commitments. (Note that this differs

> from regular FROST which just requires an authenticated channel.)

Reviewer: In Zcash the message being signed is just the sighash. How do the signers get sufficient information to understand what payment they are authorizing?

Note that the pre-signed transaction is not sufficient either; the signers need at least an opening of the value commitment correspondimg to the amount they're authorizing to be spent, and a decryption of the note ciphertexts so that they know where the funds are being sent to.

It could be argued that the details are outside the scope of this ZIP (since that will be defined by the protocol used by a given hardware wallet or similar authorizing device) but pointing out the security requirement is definitely in scope.

Signature Share Verification and Aggregation
''''''''''''''''''''''''''''''''''''''''''''

The `aggregate` function is changed to incorporate the randomizer as follows: ::

  Inputs:
  - commitment_list =
      [(j, hiding_nonce_commitment_j, binding_nonce_commitment_j), ...], a
    list of commitments issued in Round 1 by each participant, where each element
    in the list indicates a NonZeroScalar identifier j and two commitment
    Element values (hiding_nonce_commitment_j, binding_nonce_commitment_j).
    This list MUST be sorted in ascending order by identifier.
  - msg, the message to be signed, a byte string.
  - sig_shares, a set of signature shares z_i, Scalar values, for each participant,
    of length NUM_PARTICIPANTS, where MIN_PARTICIPANTS <= NUM_PARTICIPANTS <= MAX_PARTICIPANTS.
  - group_public_key, public key corresponding to the group signing key,
  - randomizer, the randomizer Scalar.

  Outputs:
  - (R, z), a Schnorr signature consisting of an Element R and Scalar z.
  - randomized_group_public_key, the randomized group public key

  def aggregate(commitment_list, msg, sig_shares, group_public_key, randomizer):
    # Compute the binding factors

>     binding_factor_list = compute_binding_factors(commitment_list, msg)

Reviewer: This needs to compute and pass in `randomizer_point`.

    # Compute the group commitment
    group_commitment = compute_group_commitment(commitment_list, binding_factor_list)

    # Compute the challenge

>     randomized_group_public_key = group_public_key + G * randomizer

Reviewer: ```suggestion
    randomized_group_public_key = group_public_key + randomizer_point
```
(We've already computed `randomizer_point` as per my comment above.)

Reviewer 3: impossible,

                             group_public_key, msg, randomizer_point):
    # Compute the randomized group public key
    randomized_group_public_key = group_public_key + randomizer_point

    # Compute the binding factors
    binding_factor_list = compute_binding_factors(commitment_list, msg, randomizer_point)
    binding_factor = binding_factor_for_participant(binding_factor_list, identifier)

    # Compute the group commitment
    group_commitment = compute_group_commitment(commitment_list, binding_factor_list)

    # Compute the commitment share
    (hiding_nonce_commitment, binding_nonce_commitment) = comm_i
    comm_share = hiding_nonce_commitment + G.ScalarMult(binding_nonce_commitment, binding_factor)

    # Compute the challenge
    challenge = compute_challenge(group_commitment, randomized_group_public_key, msg)

    # Compute the interpolating value
    participant_list = participants_from_commitment_list(commitment_list)
    lambda_i = derive_interpolating_value(identifier, participant_list)

    # Compute relation values
    l = G.ScalarBaseMult(sig_share_i)
    r = comm_share + G.ScalarMult(PK_i, challenge * lambda_i)

    return l == r

> 

Reviewer: Add the rationale / security analysis section here. (If there is separate rationale for the ciphersuites, add it in those sections.)

Ciphersuites
------------

FROST(Jubjub, BLAKE2b-512)

> '''''''''''''''''''''''''''''

Reviewer: ```suggestion
''''''''''''''''''''''''''
```

This ciphersuite uses Jubjub for the Group and BLAKE2b-512 for the Hash function `H`
meant to produce signatures indistinguishable from RedJubjub Sapling Spend
Authorization Signatures as specified in [#protocol-concretespendauthsig]_.

> - Group: Jubjub [#protocol-jubjub]_

Reviewer: The FROST spec's notion of a group includes a base point, which in this case is $\mathcal{G}^\mathsf{Sapling}$ defined in 5.4.7.1.

>   - Order: 6554484396890773809930967563523245729705921265872317281365359162392183254199 (see [#protocol-jubjub]_)

Reviewer: There's no need to repeat this value; just specify that it is $r_\mathbb{J}$ as defined in `[#protocol-jubjub]_`.

Author: Done in f72d8cd

  - Identity: as defined in [#protocol-jubjub]_
  - RandomScalar(): Implemented by returning a uniformly random Scalar in the range
    \[0, `G.Order()` - 1\]. Refer to {{frost-randomscalar}} for implementation guidance.
  - SerializeElement(P): Implemented as :math:`\mathsf{repr}_\mathbb{J}(P)` as defined in [#protocol-jubjub]_
  - DeserializeElement(P): Implemented as :math:`\mathsf{abst}_\mathbb{J}(P)` as defined in [#protocol-jubjub]_,
    failing if :math:`\bot` is returned. Additionally, this function validates that the resulting

>     element is not the group identity element, returning an error if the check fails.

Reviewer: Assuming that there is no intended distinction between "failing" and "returning an error", use the same wording.

Author: Done in f72d8cd

  - SerializeScalar: Implemented by outputting the little-endian 32-byte encoding
    of the Scalar value.
  - DeserializeScalar: Implemented by attempting to deserialize a Scalar from a
    little-endian 32-byte string. This function can fail if the input does not
    represent a Scalar in the range \[0, `G.Order()` - 1\].

- Hash (`H`): BLAKE2b-512 [#BLAKE]_ (BLAKE2b with 512-bit output and 16-byte personalization string),
  and Nh = 64.

  - H1(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubR", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer

>     modulo L = 6554484396890773809930967563523245729705921265872317281365359162392183254199.

Reviewer: ```suggestion
    modulo L = `G.Order()`.
```

  - H2(m): Implemented by computing BLAKE2b-512("Zcash_RedJubjubH", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo L = 6554484396890773809930967563523245729705921265872317281365359162392183254199.
    (This is equivalent to :math:`\mathsf{H}^\circledast(m)`, as defined in

>     [#protocol-concretereddsa]_ parametrized with [#protocol-jubjub]_.)

Reviewer: $\mathsf{H}^\circledast$ is defined in terms of $\mathsf{H}$ and $r_\mathbb{G}$ (`G.Order()`), and to be pedantic, $\mathsf{H}$ is defined by the $\mathsf{RedJubjub}$ specialisation in 5.4.7, not by the "represented group".

Author: Done in f72d8cd

  - H3(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubN", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo L = 6554484396890773809930967563523245729705921265872317281365359162392183254199.
  - H4(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubM", m)
  - H5(m): Implemented by computing BLAKE2b-512("FROST_RedJubjubC", m)

> 

Reviewer: The FROST spec [says](https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#section-6-4):
> Each ciphersuite includes explicit instructions for verifying signatures produced by FROST. Note that these instructions are equivalent to those produced by a single signer.

Add these instructions for both ciphersuites (delegating to section 5.4.7.1 of the protocol spec).

Author: Done in bc16e53

FROST(Pallas, BLAKE2b-512)

> '''''''''''''''''''''''''''''

Reviewer: ```suggestion
''''''''''''''''''''''''''
```

This ciphersuite uses Pallas for the Group and BLAKE2b-512 for the Hash function `H`
meant to produce signatures indistinguishable from RedPallas Orchard Spend
Authorization Signatures as specified in [#protocol-concretespendauthsig]_.

> - Group: Pallas [#protocol-pallasandvesta]_

Reviewer: The FROST spec's notion of a group includes a base point, which in this case is $\mathcal{G}^{Orchard}$ defined in 5.4.7.1.

Author: Done in f72d8cd

>   - Order: 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001 (see [#protocol-pallasandvesta]_)

Reviewer: Similarly, don't repeat this value.

Author: Done in f72d8cd

    element is not the group identity element, returning an error if the check fails.
  - SerializeScalar: Implemented by outputting the little-endian 32-byte encoding
    of the Scalar value.
  - DeserializeScalar: Implemented by attempting to deserialize a Scalar from a
    little-endian 32-byte string. This function can fail if the input does not
    represent a Scalar in the range \[0, `G.Order()` - 1\].

- Hash (`H`): BLAKE2b-512 [#BLAKE]_ (BLAKE2b with 512-bit output and 16-byte personalization string),
  and Nh = 64.

  - H1(m): Implemented by computing BLAKE2b-512("FROST_RedPallasR", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo L = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001.
  - H2(m): Implemented by computing BLAKE2b-512("Zcash_RedPallasH", m), interpreting
    the 64 bytes as a little-endian integer, and reducing the resulting integer
    modulo L = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001.
    (This is equivalent to :math:`\mathsf{H}^\circledast(m)`, as defined in

>     [#protocol-concretereddsa]_ parametrized with [#protocol-pallasandvesta]_.)

Reviewer: :math:`\mathsf{H}^\circledast` is defined in terms of :math:`\mathsf{H}` and `r_\mathbb{G}` (`G.Order()`), and to be pedantic, :math:`\mathsf{H}` is defined by the :math:`RedPallas` specialisation in 5.4.7.

Author: Done in f72d8cd

.. [#FROST] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html>`_
.. [#frost-protocol] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Section 5: Two-Round FROST Signing Protocol <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#section-5>`_
.. [#frost-removingcoordinator] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Section 7.3: Removing the Coordinator Role <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#section-7.3>`_
.. [#frost-primeordergroup] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Section 3.1: Prime-Order Group <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#section-3.1>`_
.. [#frost-tdkg] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Appendix B: Trusted Dealer Key Generation <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#appendix-B>`_
.. [#frost-randomscalar] `Draft RFC: Two-Round Threshold Schnorr Signatures with FROST. Appendix C: Random Scalar Generation <https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-08.html#appendix-C>`_

> .. [#protocol-concretereddsa] `Zcash Protocol Specification, Version 2022.3.4 [NU5]. Section 5.4.7: RedDSA, RedJubjub, and RedPallas <https://protocol/protocol.pdf#concretereddsa>`_

Reviewer: TODO(ZIP Editors): update these to the latest version of the spec before publishing.

::

  ZIP: 312
  Title: FROST for Spend Authorization Signatures
  Owners: Conrado Gouvea <conrado@zfnd.org>
          Chelsea Komlo <ckomlo@uwaterloo.ca>
          Deirdre Connolly <deirdre@zfnd.org>
  Status: Draft
  Category: Wallet
  Created: 2022-08-dd
  License: MIT
  Discussions-To: <https://github.com/zcash/zips/issues/382>
  Pull-Request: <https://github.com/zcash/zips/pull/662>

Terminology
===========

{Edit this to reflect the key words that are actually used.}
The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to
be interpreted as described in RFC 2119. [#RFC2119]_

The terms below are to be interpreted as follows:

> Unlinkability
>   The property of statistical independence of signatures from the signers' long-term keys, ensuring that
>   (for perfectly uniform generation of Randomizers and no leakage of metadata) it is impossible to determine whether two transactions were generated by the same
>   party.

Reviewer 4: ```suggestion
Unlinkability
  The property of statistical independence of signatures from the
  signers' long-term keys, ensuring that (for perfectly uniform
  generation of Randomizers and no leakage of metadata) it is
  impossible to determine whether two transactions were generated
  by the same party.
```

the :math:`\mathsf{ask}` is usually derived from the spending key :math:`\mathsf{sk}`,
though that is not required. Doing so might require a trusted dealer key generation
process as detailed in [#frost-tdkg]_ (as opposed to distributed key generation).

Randomizer Generation
'''''''''''''''''''''

A new helper function is defined, which computes :math:`\mathsf{RedDSA.GenRandom}`:

::

  randomizer_generate():

  Inputs:
  - None

  Outputs: randomizer, a Scalar

  def randomizer_generate():
    randomizer_input = random_bytes(64)
    return H3(randomizer_input)

Binding Factor Computation
''''''''''''''''''''''''''

The `compute_binding_factors` function is changed to receive the `randomizer_point`
as follows: ::

  Inputs:
  - commitment_list = [(i, hiding_nonce_commitment_i, binding_nonce_commitment_i), ...],
    a list of commitments issued by each participant, where each element in the list
    indicates a NonZeroScalar identifier i and two commitment Element values
    (hiding_nonce_commitment_i, binding_nonce_commitment_i). This list MUST be sorted
    in ascending order by identifier.
  - msg, the message to be signed.

>   - randomizer_point, an element in G.

Reviewer 4: Should we call this `randomizer_commitment`? It may be an elliptic curve point, or it may be a group element abstraction, but its function is as a commitment to the actual randomizer, a secret scalar

Author: Good point (:rofl:), changed in 7bffa04

      rho_input = rho_input_prefix || G.SerializeScalar(identifier)
      binding_factor = H1(rho_input)
      binding_factor_list.append((identifier, binding_factor))
    return binding_factor_list

Round One - Commitment
''''''''''''''''''''''

Roune One is exactly the same as specified #[FROST]_. But for context, it
involves these steps:

- Each signer generates nonces and their corresponding public commitments.
  A nonce is a pair of Scalar values, and a commitment is a pair of Element values.
- The nonces are stored locally by the signer and kept private for use in the second round.
- The commitments are sent to the Coordinator.

Round Two - Signature Share Generation
''''''''''''''''''''''''''''''''''''''

In Round Two, the Coordinator generates a random scalar `randomizer` by calling
`randomizer_generate`. Then it computes `randomizer_point = G.ScalarBaseMult(randomizer)`
and sends it to each signer, over a confidential and authenticated channel,
along with the message and the set of signing commitments. (Note that this differs
from regular FROST which just requires an authenticated channel.)

> In Zcash, the message that needs to be signed is actually the SIGHASH
> transaction hash, which does not convey enough information for the signers to
> decide if they want to authorize the transaction or not. Therefore, in practice,
> more data is needed to be sent (over the same encrypted, authenticated channel) from the Coordinator to the signers, possibly the
> transaction itself, openings of value commitments, decryption of note
> ciphertexts, etc.; and the signers must check that the given SIGHASH matches the
> data sent from the Coordinator, or compute the SIGHASH themselves from that
> data. However, the specific mechanism for that process is outside the scope of
> this ZIP.

Reviewer 4: ```suggestion
In Zcash, the message that needs to be signed is actually the SIGHASH
transaction hash, which does not convey enough information for the signers to
decide if they want to authorize the transaction or not. Therefore, in practice,
more data is needed to be sent (over the same encrypted, authenticated channel) 
from the Coordinator to the signers, possibly the transaction itself, openings of 
value commitments, decryption of note ciphertexts, etc.; and the signers must check 
that the given SIGHASH matches the data sent from the Coordinator, or compute the SIGHASH themselves from that data. However, the specific mechanism for that process 
is outside the scope of this ZIP.
```

split the spend authorization private key :math:`\mathsf{ask}` among multiple
possible signers. This means that the proof generation will still be performed
by a single participant, likely the one that created the transaction in the first
place. Note that this user already controls the privacy of the transaction since
they are responsible for creating the proof.

This fits well into the "Coordinator" role from the FROST specification [#frost-protocol]_.
The Coordinator is responsible for sending the message to be signed to all participants,
and to aggregate the signature shares.

With those considerations in mind, the threat model considered in this ZIP is:

- The Coordinator is trusted with the privacy of the transaction (which includes
  the unlinkability property). A rogue Coordinator will be able to break
  unlinkability and privacy, but should not be able to create signed transactions
  without the approval of `MIN_SIGNERS` participants, as specified in FROST.
- All key share holders are also trusted with the privacy of the transaction,
  thus a rogue key share holder will be able to break its privacy and unlinkability.

Non-requirements
================

> - This ZIP does not support removing the Coordinator role, as described in #[frost-removingcoordinator]_.
> - This ZIP does not prevent key share holders from linking the signing operation to a transaction in the blockchain.
> - Like the FROST specification [#FROST], this ZIP does not specify a key generation
>   procedure; but refer to that specification for guidelines.
> - Network privacy is not in scope for this ZIP, and must be obtained with other
>   tools if desired.

Reviewer 4: ```suggestion
- This ZIP does not support removing the Coordinator role, as described in
  #[frost-removingcoordinator]_.
- This ZIP does not prevent key share holders from linking the signing operation to a
  transaction in the blockchain.
- Like the FROST specification [#FROST], this ZIP does not specify a key generation
  procedure; but refer to that specification for guidelines.
- Network privacy is not in scope for this ZIP, and must be obtained with other
  tools if desired.
```

SIGHASH themselves from that data. However, the specific mechanism for that process
is outside the scope of this ZIP.

The `sign` function is changed to receive `randomizer_commitment` and incorporate it
into the computation of the binding factor. It is specified as the following: ::

  Inputs:
  - identifier, identifier i of the participant, a NonZeroScalar.
  - sk_i, Signer secret key share, a Scalar.
  - group_public_key, public key corresponding to the group signing key,
    an Element.
  - nonce_i, pair of Scalar values (hiding_nonce, binding_nonce) generated in
    round one.
  - msg, the message to be signed, a byte string.
  - commitment_list =
      [(j, hiding_nonce_commitment_j, binding_nonce_commitment_j), ...], a
    list of commitments issued in Round 1 by each participant and sent by the Coordinator.
    Each element in the list indicates a NonZeroScalar identifier j and two commitment
    Element values (hiding_nonce_commitment_j, binding_nonce_commitment_j).
    This list MUST be sorted in ascending order by identifier.
  - randomizer_commitment, an element in G (sent by the Coordinator).

  Outputs:
  - sig_share, a signature share, a Scalar.

>   def sign(identifier, sk_i, group_public_key, nonce_i, msg, commitment_list):

Author: (comment to myself to not forget it when I go back to this...)

add `randomizer_commitment`

Author: not needed anymore

This proposal adapts FROST [#FROST]_, a threshold signature scheme,
to make it unlinkable, which is a requirement for its use in the Zcash protocol.
The adapted scheme generates signatures compatible with spend authorization

> signatures in the Zcash protocol, for the Sapling and Orchard shielded pools.

Reviewer: ```suggestion
signatures in the Sapling and Orchard shielded protocols as deployed in Zcash.

along with the message and the set of signing commitments. (Note that this differs
from regular FROST which just requires an authenticated channel.)

In Zcash, the message that needs to be signed is actually the SIGHASH

> transaction hash, which is does not convey enough information for the signers to

Reviewer: ```suggestion
transaction hash, which does not convey enough information for the signers to
```

decide if they want to authorize the transaction or not. Therefore, in practice,

> more data is needed to be sent from the Coordinator to the signers, possibly the

Reviewer: ```suggestion
more data is needed to be sent (over the same encrypted, authenticated channel) from the Coordinator to the signers, possibly the
```
